x/datadeal/keeper/deal.go

```diff
@@ -230,3 +233,89 @@ func (k Keeper) GetAllDataSaleList(ctx sdk.Context) ([]types.DataSale, error) {
 
 	return dataSales, nil
 }
+
+func (k Keeper) VoteDataDelivery(ctx sdk.Context, vote *types.DataDeliveryVote, signature []byte) error {
+	if err := vote.ValidateBasic(); err != nil {
+		return sdkerrors.Wrap(types.ErrDataDeliveryVote, err.Error())
+	}
+	if !k.verifyVoteSignature(ctx, vote, signature) {
+		return sdkerrors.Wrap(oracletypes.ErrDetectionMaliciousBehavior, "")
+	}
+
+	// Check if the dataSale vote status
+	if err := k.validateDataDeliveryVote(ctx, vote); err != nil {
+		return sdkerrors.Wrap(types.ErrDataDeliveryVote, err.Error())
+	}
+
+	// When all validations pass, the vote is saved.
+	if err := k.SetDataDeliveryVote(ctx, vote); err != nil {
+		return sdkerrors.Wrap(types.ErrDataDeliveryVote, err.Error())
+	}
+
+	return nil
+}
+
+func (k Keeper) verifyVoteSignature(ctx sdk.Context, vote *types.DataDeliveryVote, signature []byte) bool {
+	voteBz := k.cdc.MustMarshal(vote)
+
+	oraclePubKeyBz := k.oracleKeeper.GetParams(ctx).MustDecodeOraclePublicKey()
+	return secp256k1.PubKey(oraclePubKeyBz).VerifySignature(voteBz, signature)
+}
+
+func (k Keeper) validateDataDeliveryVote(ctx sdk.Context, vote *types.DataDeliveryVote) error {
+	oracle, err := k.oracleKeeper.GetOracle(ctx, vote.VoterAddress)
+	if err != nil {
+		return err
+	}
+	if oracle.Status != oracletypes.ORACLE_STATUS_ACTIVE {
+		return fmt.Errorf("this oracle is not in 'ACTIVE' state")
```

Comment:
There is the same error in #438. It would be also good to make a sentinel error for reuse.

Reply:
I fixed this part, but please check if it is the correct.